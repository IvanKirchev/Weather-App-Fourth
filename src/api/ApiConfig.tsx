export default {
    openweatherUrl: 'https://api.openweathermap.org',
    // TODO don't store secrets in source control
    openWeatherAppID: '3045bd446337c6e9aa7ef0719614442d'
}