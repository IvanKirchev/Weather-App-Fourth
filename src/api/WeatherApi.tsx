import Forecast from "../components/Forecast";

interface IForecast {
    city: ICity
    list: IHourForecast[]
    days: IDayForecast[]
}

interface IDayForecast {
    list: IHourForecast[]
    id: string
    name: string,
    averageTemp: number
    order: number
    forecastByHours: IHourForecast[]
    icon: string,
    minTemp: number,
    maxTemp: number,
    weatherDescription: string
}

interface IHourForecast {
    dt: number
    main: IMainForecastData
    weather: IWeatherDescription[]
    wind: IWind
    visibility: number
}

interface IMainForecastData {
        temp: number
        feels_like: number
        temp_min: number
        temp_max: number
        pressure: number
        sea_level: number
        grnd_level: number
        humidity: number
        temp_kf: number
}

interface IWeatherDescription {
    id: number
    main: string
    description: string
    icon: string
}

interface IWind {
    speed: number
    deg: number
    gust: number
}

interface ICity {
    country: string
    id: number
    name: string
    population: number
    sunrise: number
    sunset: number
    timezone: number
}

export type { IForecast, IDayForecast, IHourForecast, IMainForecastData, IWeatherDescription, IWind };
