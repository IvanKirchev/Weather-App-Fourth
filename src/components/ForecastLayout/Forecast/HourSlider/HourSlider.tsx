import { Box, Slider } from "@mui/material";
import react, { Dispatch, SetStateAction } from "react";
import { IDayForecast, IHourForecast } from "../../../../api/WeatherApi";

interface HourSliderProps {
  day: IDayForecast | null;
  selectedHour: IHourForecast | null;
  setSelectedHour: Dispatch<SetStateAction<IHourForecast | null>>;
  selectedHourIndex: number;
  setSelectedHourIndex: Dispatch<SetStateAction<number>>;
}

function HourSlider(props: HourSliderProps) {
  function getHourByIndex(value: number) {
    if (props.day) {
      let date = new Date(props.day?.forecastByHours[value].dt * 1000);
      return date.toLocaleTimeString([], {
        hour: "2-digit",
        minute: "2-digit",
        timeZoneName: "short",
      });
    } else {
      return "";
    }
  }

  const handleChange = (event: Event, newValue: number | number[]) => {
    if (typeof newValue === "number") {
      if (props.day?.forecastByHours[newValue] != undefined) {
        let selected = props.day?.forecastByHours[newValue];
        props.setSelectedHour(selected);
        // When the slider is changed (value change or rerender) set the current value to the last avaible index if the new value is not available
        //console.log(newValue)
        props.setSelectedHourIndex(
          props.day.forecastByHours.length > newValue
            ? newValue
            : props.day.forecastByHours.length - 1
        );
      }
    }
  };

  const marks = props.day ? props.day.forecastByHours.map((hour, index) => {
    let newIndex = index || 0;

    return {
      value: index,
      label: getHourByIndex(newIndex),
    };
  }) : [];

  function getMaxSliderValue(): number {
    if(props.day && props.day.forecastByHours.length - 1 === 0) {
      return 1
    } else if(props.day) {
      return props.day.forecastByHours.length - 1
    } 

    return 0;
  }

  return (
    <>
      {props.day ? (
        <div>
          <Box sx={{ width: "41rem", marginLeft: "34%" }}>
            <Slider
              aria-label="Temperature"
              value={props.selectedHourIndex}
              valueLabelDisplay="off"
              step={1}
              marks={marks}
              min={0}
              max={getMaxSliderValue()}
              onChange={handleChange}
            />
          </Box>
        </div>
      ) : null}
    </>
  );
}

export default HourSlider;
