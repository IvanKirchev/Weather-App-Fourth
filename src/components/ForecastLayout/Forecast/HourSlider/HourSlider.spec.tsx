import React from "react";
import renderer from "react-test-renderer";
import HourSlider from "./HourSlider";
import enzyme from "enzyme";
import { IDayForecast, IHourForecast } from "../../../../api/WeatherApi";

const mockHour = {
  dt: 16372368,
  main: {
    temp: 13,
    feels_like: 13,
    temp_min: 13,
    temp_max: 13,
    pressure: 13,
    sea_level: 13,
    grnd_level: 13,
    humidity: 13,
    temp_kf: 13,
  },
  weather: [
    {
      id: 13,
      main: "13",
      description: "Nice weather",
      icon: "icon url",
    },
  ],
  wind: {
    speed: 13,
    deg: 13,
    gust: 14,
  },
  visibility: 1000,
};

const mockDayForecast: IDayForecast = {
  id: 1637236800000,
  name: "Thursday",
  order: 0,
  icon: "04d",
  forecastByHours: [mockHour],
  weatherDescription: "broken clouds",
  averageTemp: 7.445,
  minTemp: 0,
  maxTemp: 0,
};

it("Should render HourSlider component correctly", () => {
  const component = renderer.create(
    <HourSlider day={mockDayForecast} selectedHour={mockHour} setSelectedHour={() => {}} setSelectedHourIndex={() => {}} selectedHourIndex={0}/>
  );

  expect(component).toMatchSnapshot()
});
