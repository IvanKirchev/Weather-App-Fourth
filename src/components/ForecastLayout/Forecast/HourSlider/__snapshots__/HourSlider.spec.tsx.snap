// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render HourSlider component correctly 1`] = `
<div>
  <div
    className="MuiBox-root css-15e9aw9"
  >
    <span
      className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root MuiSlider-marked css-1aznpnh-MuiSlider-root"
      onMouseDown={[Function]}
    >
      <span
        className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
      />
      <span
        className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
        data-index={0}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <span
        aria-hidden={true}
        className="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
        data-index={0}
        style={
          Object {
            "left": "0%",
          }
        }
      >
        01:52 PM GMT+2
      </span>
      <span
        className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
        data-index={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "0%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-label="Temperature"
          aria-orientation="horizontal"
          aria-valuemax={1}
          aria-valuemin={0}
          aria-valuenow={0}
          data-index={0}
          disabled={false}
          max={1}
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={0}
        />
      </span>
    </span>
  </div>
</div>
`;
